routing-configurer/src/main/java/no/vegvesen/ixn/federation/ServiceProviderRouter.java

```diff
@@ -187,45 +187,32 @@ private void optionallyCreateQueue(String queueName, String serviceProviderName,
         }
     }
 
-    public ServiceProvider syncPrivateChannels(ServiceProvider serviceProvider, QpidDelta delta) {
-        if (!serviceProvider.getPrivateChannels().isEmpty()) {
-            Set<PrivateChannel> privateChannels = new HashSet<>();
-            privateChannels.addAll(serviceProvider.getPrivateChannels());
-            if (!privateChannels.isEmpty()) {
-                syncPrivateChannelsWithQpid(privateChannels, serviceProvider.getName(), delta);
-                Set<PrivateChannel> privateChannelsToRemove = privateChannels
-                        .stream()
-                        .filter(s -> s.getStatus().equals(PrivateChannelStatus.TEAR_DOWN))
-                        .collect(Collectors.toSet());
-
-                privateChannels.removeAll(privateChannelsToRemove);
-                serviceProvider.setPrivateChannels(privateChannels);
-            }
-            serviceProvider = repository.save(serviceProvider);
+    public void syncPrivateChannels(ServiceProvider serviceProvider, QpidDelta delta) {
+        List<PrivateChannel> privateChannelList = privateChannelRepository.findAllByServiceProviderName(serviceProvider.getName());
+        if (!privateChannelList.isEmpty()) {
+            syncPrivateChannelsWithQpid(privateChannelList, serviceProvider.getName(), delta);
+            privateChannelList.stream().filter((a) -> a.getStatus().equals(PrivateChannelStatus.TEAR_DOWN)).forEach(privateChannelRepository::delete);
         }
-        return serviceProvider;
+
     }
 
-    private void syncPrivateChannelsWithQpid(Set<PrivateChannel> privateChannels, String name, QpidDelta delta) {
-        Set<PrivateChannel> privateChannelsWithStatusCreated = privateChannels
-                .stream()
-                .filter(s -> s.getStatus().equals(PrivateChannelStatus.CREATED))
-                .collect(Collectors.toSet());
+    private void syncPrivateChannelsWithQpid(List<PrivateChannel> privateChannels, String name, QpidDelta delta) {
 
-        GroupMember groupMember = qpidClient.getGroupMember(name,CLIENTS_PRIVATE_CHANNELS_GROUP_NAME);
+        List<PrivateChannel> privateChannelsWithStatusCreated = privateChannelRepository.findAllByStatusAndServiceProviderName(PrivateChannelStatus.CREATED, name);
+        GroupMember groupMember = qpidClient.getGroupMember(name, CLIENTS_PRIVATE_CHANNELS_GROUP_NAME);
         if (groupMember == null) {
             qpidClient.addMemberToGroup(name, CLIENTS_PRIVATE_CHANNELS_GROUP_NAME);
             logger.debug("Adding member {} to group {}", name, CLIENTS_PRIVATE_CHANNELS_GROUP_NAME);
         }
 
-        for(PrivateChannel privateChannel : privateChannels) {
+        for (PrivateChannel privateChannel : privateChannels) {
```

Comment:
This should be moved down into the if REQUESTED. Should not have any instance of private channel being CREATED and not having a name, and if it's TEAR_DOWN and the name is removed, this will lead to problems.